### DISCUSSION TITLE
Improved methods for git authorization

### DISCUSSION THREAD

**Discussion by ras0219-msft** [microsoft/vcpkg#17272] [Ideas]
In addition to `vcpkg_from_git()`, registries also now directly use the git command line tools to fetch remote resources. Some of these resources may need to be protected from anonymous access, so it's important to be able to provide credentials.

### Pre-seed git credentials

_This method is newly proposed in this discussion, please chime in with feedback if this works for you!_

Based on my current understanding, we can pre-seed git credentials via `git credential approve`:
Powershell:
```powershell
"url=https://github.com`npath=Microsoft/vcpkg`nusername=unused`npassword=$MY_PAT`n" | git credential approve
```
Bash:
```sh
echo "url=https://github.com"$'\n'"path=Microsoft/vcpkg"$'\n'"username=unused"$'\n'"password=$MY_PAT"$'\n' | git credential approve
```

For systems which need bearer auth, you can use `git config`:
**Note: you must make these config changes with `--global`**
```
git config --global --unset-all http.<uri>.extraheader
git config --global http.<uri>.extraheader "AUTHORIZATION: bearer <System_AccessToken>"
```
The `<uri>` can be filled in with a variety of options, documented in https://git-scm.com/docs/git-config#Documentation/git-config.txt-httplturlgt. For example, `https://dev.azure.com/MYORG/`.

(Original Source: https://github.com/Microsoft/azure-pipelines-agent/issues/1601#issuecomment-394511048).

**Note for Azure DevOps users:** You may need to enable access via Job authorization scope https://docs.microsoft.com/en-us/azure/devops/pipelines/process/access-tokens?view=azure-devops&tabs=yaml#job-authorization-scope. You may need to "reference" the repo in your yaml via:

```yaml
resources: 
  repositories:
    - repository: <FRIENDLYNAME>
      type: git
      name: <ORG>/<REPO>
      tag: tags/<TAG>

...

jobs:
 - job: Build
   uses:
     repositories: [<FRIENDLYNAME>]
```

### Pass credentials in an environment variable

Using `VCPKG_KEEP_ENV_VARS` or `VCPKG_ENV_PASSTHROUGH_UNTRACKED`, we can smuggle credential info via another var like `MY_TOKEN`.
```sh
export VCPKG_KEEP_ENV_VARS=MY_TOKEN
export MY_TOKEN=abc123
```
Then, this can be used in private ports:
```cmake
# some/private/portfile.cmake
if(DEFINED ENV{MY_TOKEN})
    set(ENV{MY_TOKEN} "$ENV{MY_TOKEN}@")
endif()
vcpkg_from_git(
    URLS "https://$ENV{MY_TOKEN}host.com/normal/url/path"
    ...
)
```
```cmake
# some/other/private/portfile.cmake
vcpkg_from_github(
    AUTHORIZATION_TOKEN "$ENV{MY_TOKEN}"
)
```

Drawback: This doesn't work for git registries :(

+@donny-dont as the author of https://github.com/microsoft/vcpkg/pull/5719

**Comment by ras0219-msft** [comment]
See also https://github.com/microsoft/vcpkg/discussions/17720#discussioncomment-708467 which discusses a similar solution

**Comment by simon-winter** [comment]
while the git credential solutions did not work for me, i could pass the credentials in the env variable like you described. Actually one can do the same for the repository. So when you have a private repository and want to access it from your local repo, you cann pass a PAT infront of the url in the `vcpkg-configuration.json` like so:
```
{
  "registries": [
    {
      "kind": "git",
      "repository": "https://564kjh57someGeneratedSuperlongToken876jdfshgdfgn@dev.azure.com/yourCompany/someRepo/_git/vcpkg-registry",
      "packages": [ "yourPrivateLib" ],
      "baseline": "theDesiredCommitIdOfTheVcpkg-registry"
    }
  ]
}
```
this saves obviously the credentials in your project, so it cannot have less security then your registry but in my case its a valid solution, as i can access my repo via my company login, but i crosscompile to WSL and loose my access rights, so having them inside the repo fixes this issue. As the build files never leave the company it should be fine

**Comment by thomthom** [comment]
I'm trying to make a private vcpkg package. It's using `vcpkg_from_github`.

Is `vcpkg_from_github` using the system default git instance under the hood? Or it's own instance?

From Windows Terminal I'm able to use git CLI to push/pull from the private repo just fine. But when vcpkg tries to download it I get 404 error. (I double checked the URL via the browser, and the URL is valid when you are logged in.)

```
[DEBUG] Downloading https://github.com/example/example-lib/archive/3a249ffbb1b405ef3035dd8f382564729d9c9ba9.tar.gz
[DEBUG] D:\a\_work\1\s\src\vcpkg\base\downloads.cpp(656)
Error: Failed to download from mirror set:
https://github.com/example/example-lib/archive/3a249ffbb1b405ef3035dd8f382564729d9c9ba9.tar.gz: failed: status code 404
```

I did try to use `git credential approve` for the sake of it, but that didn't seem to work either.

Oh wait, I just noticed the log says it's downloading a `.tar.gz` file. This means `vcpkg_from_github` not using `git`? Do I have to switch to `vcpkg_from_git` for this to work?

**Reply by ras0219-msft** [reply]
Yep, exactly your last paragraph. `vcpkg_from_github()` transforms things into a .tar.gz to maximize efficiency for the build. If you want to do a private repo using git auth, you'll need to use `vcpkg_from_git()`.

**Reply by thomthom** [reply]
Thank you. It wasn't entirely clear from the documentation how things worked under the hood. The guides for creating your own private registry and package is pretty good, but I wished it was complimented with examples of how "normal git methods" for accessing worked in regard to the various methods of obtaining the source. Especially for setting up within CI setting. This thread is a good start, but it wasn't easy to come across. I'll be trying it out shortly with GitHub CI once I get my registry and package working.

**Reply by ras0219-msft** [reply]
I've recently added http://github.com/Microsoft/vcpkg/tree/master/docs/users/authentication.md which hopefully will make this information easier to find :)

**Reply by sketch34** [reply]
@ras0219-msft thank you! I'm attempting to set this up, but I'm a little lost at the `$MY_PAT` in the instructions you link to above. Is this some bash variable that you are defining elsewhere?

**Comment by HofmeisterAn** [comment]
Does anyone know if global, pre-seeded git credentials may conflict concurrent builds? If a pipeline sets (unsets) the extraheader in the global git config like `http.https://Foo@dev.azure.com/Foo.extraheader` and another build checks out:

```
git config --get-all http.https://Foo@dev.azure.com/Foo/ProjectName/_git/GitRepository.extraheader
git -c http.extraheader="AUTHORIZATION: bearer ***" fetch --force --tags --prune --progress --no-recurse-submodules origin
```
Does it work? I did some locale tests, they look fine so far. Is there anything I need to keep an eye on?
